TCUP: Typing and Characterization using Proteomics
==================================================
TCUP uses peptides generated by mass spectrometry proteomics to:

* identify and estimate the taxonomic composition of a microbial sample.
* detect expressed antibiotic resistance proteins.

TCUP does not require a priori information about the contents of a sample and
is suitable for analysis of both pure cultures, mixed samples, and clinical
samples.  Have a look in the online `documentation`_ to learn how to use
TCUP.

.. _documentation: https://tcup.readthedocs.org


About
*****
:Authors: Fredrik Boulund
:Contact: erik.kristiansson@chalmers.se
:License: ISC


Publication
***********
If you use TCUP in your work, please cite::

   Fredrik Boulund, Roger Karlsson, Lucia Gonzales-Siles, Anna Johnning, Nahid Karami, Omar AL-Bayati, Christina Ahren, Edward R. B. Moore, and Erik Kristiansson
   TCUP: Typing and characterization of bacteria using bottom-up tandem mass spectrometry proteomics
   Mol Cell Proteomics mcp.M116.061721. First Published on April 18, 2017, doi:10.1074/mcp.M116.061721

To cite this repository, please use:

.. image:: https://zenodo.org/badge/202481367.svg
   :target: https://zenodo.org/badge/latestdoi/202481367


Installation 
************
Detailed installation instructions are available in the online
`documentation`_.

Download and install Anaconda Python 3.5. Then create a conda environment
and install TCUP into the environment using the following commands::

    $ hg clone https://bitbucket.org/chalmersmathbioinformatics/tcup
    $ conda env create -f ./tcup/conda_environment_Linux.yml
    $ source activate tcup
    $ pip install ./tcup/

This will create a conda environment called ``tcup`` that contains 
all the required dependencies, and the ``tcup`` package itself. 

Dependencies
------------
TCUP depends on the following Python packages, easily installable via
``conda`` and ``pip``.

 * `ETE Toolkit`_ (ete3)
 * `XlsxWriter`_ (xlsxwriter)

.. _XlsxWriter: http://xlsxwriter.readthedocs.org/
.. _ETE Toolkit: http://etetoolkit.org/

Running
*******
In order to run TCUP, some databases need to be prepared. Please consult the
online `documentation`_ for details on how to build these.

The input for TCUP is mappings of peptides to reference genomes in blast8
tabular format. We recommend using `BLAT`_ or `pBLAT`_ for mapping.

.. _BLAT: https://genome.ucsc.edu/FAQ/FAQblat.html
.. _pBLAT: http://icebert.github.io/pblat/
